Maintenance: Prepare migration of NSKeyedArchiver methods and move to block-based animation

wutschel opened the pull request:
## Description
<!--- Detailed info for reviewers and developers -->
Closes https://github.com/xbmc/Official-Kodi-Remote-iOS/issues/599.

This PR has four main sections:
1. Prepare migration of `NSKeyedArchiver` methods
The method `unarchiveObjectWithFile` and `archiveRootObject:toFile:` are deprecated since iOS 11. This PR refactors reading/writing persistent data and already prepares the migration to the newly introduced `unarchivedObjectOfClasses` and `archivedDataWithRootObject:requiringSecureCoding:error:`. 
2. Move duplicate animation related code to Utilities
This allows to remove code duplication and unneeded code, as well as prepares for the next third step.
3. Migrate to block-based animation
All parameters and options were taken over into the block-based animation.
4. Optimizations and bugfixes around animation in `NowPlaying`
The code was refactored to remove unneeded variables. In addition, the animation for the transition between playlist and NowPlaying view was corrected (now properly sliding out and in). As a further optimization the "demo" animation for the playlist/NowPlaying toggle button is starting earlier. This always bugged me.

## Summary for release notes
<!--- This will be shown in Testflight to end users / testers -->
<!--- Please fill in, as usually PR title isn't clear to ordinary users -->
<!--- If your changes don't modify app files, please add prefix [not app] -->
Maintenance: Prepare migration of NSKeyedArchiver methods
Maintenance: Migrate to block-based animation
Bugfix: Fix flip in/out animation for playlist/NowPlaying view

wutschel commented:
@kambala-decapitator, I am keeping this as is for the review. There are multiple small commits which should make this PR easier to review. One this is done I propose to squash all commits of section 2 and section 3 into a single commit each.

kambala-decapitator commented on XBMC Remote/ViewControllerIPad.m:
```diff
@@ -609,12 +609,14 @@ - (void)handleTcpJSONRPCChangeServerStatus:(NSNotification*)sender {
+                        options:UIViewAnimationOptionCurveEaseInOut
+                     animations:^{
+        self.nowPlayingController.songDetailsView.alpha = 0.0;
+        self.nowPlayingController.itemDescription.scrollsToTop = NO;
```
this property is not animatable

wutschel replied:
Does it harm to keep `self.nowPlayingController.itemDescription.scrollsToTop = NO;` in the animation block? Or does it need to be moved out? Before this change it was also in the animated block.

wutschel replied:
Done

kambala-decapitator replied:
no harm here, I just prefer that animation block contains only animations to make things clearer

kambala-decapitator commented on XBMC Remote/RemoteController.m:
```diff
@@ -402,21 +402,25 @@ - (void)touchesCancelled:(NSSet*)touches withEvent:(UIEvent*)event {
+                          delay:0.0
+                        options:UIViewAnimationOptionCurveEaseInOut
+                     animations:^{
+        subsInfoLabel.alpha = 0;
```
you already have utility method to animate alpha, why not using it? there're also other places like this.

wutschel replied:
> I'd squash most of the animation commits into a single one

Yes, I wanted to squash 
1. [Use the block-based animation API] up to [Use the block-based animation API (7)]2. 
2. [Move AnimTable] up to [Remove alphaButton]

> now you have utility methods for label / table / button / view, but all of these elements are views, don't see a reason to have dedicated methods that essentially do the same thing.

Hmm, I didn't look into this yet. Let me see.

> also it seems that in some places indentation broke.

Well, especially `StackScrollViewController` has completely messed up indentations. Shall I just auto-indent this in a commit at the end?

wutschel replied:
Is `UIViewAnimationOptionCurveEaseInOut` the default option?

wutschel replied:
It is `UIViewAnimationOptionCurveEaseInOut = 0 << 16`.

Nevertheless explicitly added `UIViewAnimationOptionCurveEaseInOut` as option the the helper methods.

kambala-decapitator replied:
I was thinking about second method that accepts the curve parameter, will check your solution

kambala-decapitator replied:
yeah, always this param is used, no need for a second method

kambala-decapitator commented on XBMC Remote/NowPlaying.m:
```diff
@@ -1438,18 +1440,20 @@ - (void)flipAnimButton:(UIButton*)button demo:(BOOL)demo {
                              [button setImage:image forState:UIControlStateSelected];
                          }
-                         [UIView setAnimationCurve:UIViewAnimationCurveEaseIn];
                          [UIView setAnimationTransition:anim forView:button cache:YES];
```
is this line removed in next commit?

kambala-decapitator replied:
thanks, was difficult to notice in that huge commit

kambala-decapitator commented on XBMC Remote/NowPlaying.m:
```diff
@@ -1438,18 +1440,20 @@ - (void)flipAnimButton:(UIButton*)button demo:(BOOL)demo {
-                         [UIView beginAnimations:nil context:nil];
-                         button.hidden = NO;
-                         [UIView setAnimationDuration:0.5];
-                         [UIView setAnimationDelegate:self];
```
I don't see delegate method being removed / its logic being ported. Or is there no such method?

wutschel replied:
There is nothing like this. I guess this was just superfluous.

kambala-decapitator reviewed (changes requested):
I'd squash most of the animation commits into a single one: now you have utility methods for label / table / button / view, but all of these elements are views, don't see a reason to have dedicated methods that essentially do the same thing.

also it seems that in some places indentation broke.

wutschel commented on XBMC Remote/Utilities.m:
```diff
@@ -977,4 +977,41 @@ + (void)enableDefaultController:(id<UITableViewDelegate>)viewController tableVie
+
++ (void)archivePath:(NSString*)path file:(NSString*)filename data:(id)data {
+    NSString *filePath = [path stringByAppendingPathComponent:filename];
+    NSError *error;
+    NSData *archiveData;
```
Done

wutschel commented on XBMC Remote/Utilities.m:
```diff
@@ -977,4 +977,41 @@ + (void)enableDefaultController:(id<UITableViewDelegate>)viewController tableVie
+        unarchived = [NSKeyedUnarchiver unarchiveObjectWithFile:filePath];
+    }
+    
+    if (error) {
+        return nil;
+    } else {
+        return unarchived;
+    }
```
Done

wutschel commented on XBMC Remote/Utilities.m:
```diff
@@ -977,4 +977,41 @@ + (void)enableDefaultController:(id<UITableViewDelegate>)viewController tableVie
+    id unarchived;
+    
+    if (@available(iOS 11.0, *)) {
+        unarchived = [NSKeyedUnarchiver unarchivedObjectOfClasses:[NSSet setWithArray:objectClasses]
```
True, changed it.

wutschel commented on XBMC Remote/AppDelegate.m:
```diff
@@ -384,27 +384,30 @@ - (NSDictionary*)sortmethod:(NSString*)method order:(NSString*)order ignoreartic
-        if (![fileManager1 fileExistsAtPath:self.libraryCachePath]) {
-            [fileManager1 createDirectoryAtPath:self.libraryCachePath withIntermediateDirectories:YES attributes:nil error:NULL];
+        self.libraryCachePath = [paths[0] stringByAppendingPathComponent:@"LibraryCache"];
+        if (![fileManager fileExistsAtPath:self.libraryCachePath]) {
+            [fileManager createDirectoryAtPath:self.libraryCachePath
```
Ok, removed the check -- also for the 2nd case.

wutschel commented on XBMC Remote/AppDelegate.m:
```diff
@@ -384,27 +384,30 @@ - (NSDictionary*)sortmethod:(NSString*)method order:(NSString*)order ignoreartic
-        if ([fileManager1 fileExistsAtPath:self.dataFilePath]) {
-            NSMutableArray *tempArray;
-            tempArray = [NSKeyedUnarchiver unarchiveObjectWithFile:self.dataFilePath];
+        self.dataFilePath = paths[0];
```
Agreed, done.

wutschel commented on XBMC Remote/AppDelegate.m:
```diff
@@ -384,27 +384,30 @@ - (NSDictionary*)sortmethod:(NSString*)method order:(NSString*)order ignoreartic
-//        [[NSFileManager defaultManager] removeItemAtPath:self.epgCachePath error:nil];
-        if (![fileManager1 fileExistsAtPath:self.epgCachePath]) {
-            [fileManager1 createDirectoryAtPath:self.epgCachePath withIntermediateDirectories:YES attributes:nil error:NULL];
+        if (![fileManager fileExistsAtPath:self.epgCachePath]) {
```
But it also does not harm to keep this. If you don't mind, let us not change.

kambala-decapitator replied:
sure, I don't insist

wutschel commented:
> II'd squash most of the animation commits into a single one

Yes, I wanted to squash
1. [Use the block-based animation API] up to [Use the block-based animation API (7)]2.
2. [Move AnimTable] up to [Remove alphaButton]3. 

> now you have utility methods for label / table / button / view, but all of these elements are views, don't see a reason to have dedicated methods that essentially do the same thing.

Hmm, I didn't look into this yet. Let me see.

> also it seems that in some places indentation broke.

Well, especially `StackScrollViewController` has completely messed up indentations. Shall I just auto-indent this file in a commit at the end?

wutschel commented:
Anything more I should work on? Or are we good and I can start to squash and rebase?

kambala-decapitator commented on XBMC Remote/Utilities.m:
```diff
@@ -1031,25 +1023,12 @@ + (void)AnimView:(UIView*)view AnimDuration:(NSTimeInterval)seconds Alpha:(CGFlo
-        frame = button.frame;
-        frame.origin.x = X;
-        button.frame = frame;
-        button.hidden = hiddenValue;
```
- `hidden` isn't an animatable property
- you replaced `hidden` with `alpha` (which is animatable), is it intended?

wutschel replied:
Yes, this was intended.

kambala-decapitator commented on XBMC Remote/Utilities.m:
```diff
@@ -1060,13 +1039,7 @@ + (void)alphaView:(UIView*)view AnimDuration:(NSTimeInterval)seconds Alpha:(CGFl
-                     animations:^{
-        image.alpha = alphavalue;
-        if (alphavalue) {
-            image.hidden = NO;
```
also hidden/alpha point

wutschel replied:
Yes, I was taking care if this. `alphaImage:` was only called in few places. The view handed over was either `fanartView` which never is hidden, or `jewelView` for which I added an explicit unhide before calling the unified method `alphaView:`. So, the `hidden = NO` is not required inside the helper method anymore.

kambala-decapitator commented:
squash: sounds good

indent: yes, let's do that

wutschel commented:
Squashed and rebased. I skipped the auto-indent as I forgot I already have some work prepared in the background which conflicts with this. I will do this next time I am touching this file as part of a PR.

kambala-decapitator reviewed (approved):
just to double check: this is ready to be merged, isn't it?

wutschel commented:
> just to double check: this is ready to be merged, isn't it?

Yes, it is.

kambala-decapitator commented:
@wutschel have you seen this warning in the console?

> NSSecureCoding allowed classes list contains [NSObject class], which bypasses security by allowing any Objective-C class to be implicitly decoded. Consider reducing the scope of allowed classes during decoding by listing only the classes you expect to decode, or a more specific base class than NSObject.

wutschel commented:
> I guess it can be ignored since secure coding isn't used?

I have seen this yesterday for the first time. Not sure why it was not coming up before. Maybe it only showed after I removed the App.

I do not know what this can cause. I'd rather take a deeper look. In worst case I simply provide a PR which comments the new code for 1.10 to give time to deal this properly.

kambala-decapitator commented:
I think there's no issue, but please investigate shortly to be 100% sure
